### PR TITLE
Attach source code to certain errors

### PR DESCRIPTION
**Pull request by shaobo-he-aws**
## Issue #s, if available

#977, #1335

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar language specification.

### CODE REVIEW

**Comment by john-h-kastner-aws** [comment]
Should probably also fix #948

File: cedar-policy-core/src/test_utils.rs [hunk 1]
```diff
@@ -161,11 +161,11 @@ impl<'a> ExpectedErrorMessage<'a> {
     /// `src` is the full source text (which the miette labels index into).
     /// It can be omitted only in the case where we expect no underlines.
     /// Panics if this invariant is violated.
```
Comment (cdisselkoen) [inline cedar-policy-core/src/test_utils.rs:163]: Presumably we can remove this comment?

File: cedar-policy-core/src/test_utils.rs [hunk 2]
```diff
@@ -285,15 +285,16 @@ impl<'a> ExpectedErrorMessage<'a> {
...
         let actual_num_labels = err.labels().map(|iter| iter.count()).unwrap_or(0);
         if expected_num_labels != actual_num_labels {
             return false;
         }
         if expected_num_labels != 0 {
-            let src =
-                src.expect("src can be `None` only in the case where we expect no underlines");
+            let src = err
+                .source_code()
+                .expect("src can be `None` only in the case where we expect no underlines");
```
Comment (cdisselkoen) [inline cedar-policy-core/src/test_utils.rs:297]:
```suggestion
                .expect("err.source_code() should be `Some` if we are expecting underlines");
```

File: cedar-policy-validator/src/cedar_schema/err.rs [hunk 3]
```diff
@@ -102,26 +103,27 @@ lazy_static! {
...
+        Self::from_raw_parse_error(recovery.error, src)
     }
 }
 
 impl ParseError {
     /// Extract a primary source span locating the error.
     pub fn primary_source_span(&self) -> SourceSpan {
-        let Self { err } = self;
+        let Self { err, .. } = self;
         match err {
```
Comment (cdisselkoen) [inline cedar-policy-validator/src/cedar_schema/err.rs:127]: does this work? Wondering why the existing code is written like it is
```suggestion
        match &self.err {
```

Comment (cdisselkoen) [reply cedar-policy-validator/src/cedar_schema/err.rs:127]: Also occurs a couple of other times in this file

**Review by john-h-kastner-aws — APPROVED** [review]

**Review by cdisselkoen — APPROVED** [review]
Looks great

File: cedar-policy-core/src/ast/name.rs [hunk 4]
```diff
@@ -394,11 +394,12 @@ impl std::fmt::Display for ValidSlotId {
...
 /// [`SlotId`] plus a source location
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct Slot {
     /// [`SlotId`]
     pub id: SlotId,
     /// Source location, if available
+    #[serde(skip)]
     pub loc: Option<Loc>,
 }
```
[the remark this answers is not in the document]
Comment (shaobo-he-aws) [reply cedar-policy-core/src/ast/name.rs:404]: This change is introduced by a previous revision and unnecessary. Reverted.

File: cedar-policy-core/src/parser/err.rs [hunk 5]
```diff
@@ -472,14 +473,32 @@ pub mod parse_errors {
...
     }
 
+    impl Diagnostic for ExpectedStaticPolicy {
+        fn help<'a>(&'a self) -> Option<Box<dyn Display + 'a>> {
+            Some(Box::new(
+                "try removing the template slot(s) from this policy",
+            ))
+        }
+
+        fn labels(&self) -> Option<Box<dyn Iterator<Item = miette::LabeledSpan> + '_>> {
```
[the remark this answers is not in the document]
Comment (shaobo-he-aws) [reply cedar-policy-core/src/parser/err.rs:490]: Looking into it.

[the remark this answers is not in the document]
Comment (shaobo-he-aws) [reply cedar-policy-core/src/parser/err.rs]: It should work now.

File: cedar-policy-core/Cargo.toml [hunk 6]
```diff
@@ -42,6 +42,7 @@ chrono = { version = "0.4.38", optional = true, default-features = false}
 
 # protobuf dependency
 prost = { version = "0.13", optional = true }
+educe = "0.6.0"
```
Comment (cdisselkoen) [inline cedar-policy-core/Cargo.toml:46]: is the "protobuf dependency" section the correct place to put `educe`?

File: cedar-policy-core/src/ast/policy.rs [hunk 7]
```diff
@@ -1694,7 +1702,10 @@ impl EntityReference {
     pub fn into_expr(&self, slot: SlotId) -> Expr {
         match self {
             EntityReference::EUID(euid) => Expr::val(euid.clone()),
-            EntityReference::Slot => Expr::slot(slot),
+            EntityReference::Slot(loc) => loc
+                .as_ref()
+                .map(|loc| ExprBuilder::new().with_source_loc(loc.clone()).slot(slot))
+                .unwrap_or(Expr::slot(slot)),
```
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/policy.rs:1708]: maybe could be simplified to something like
```suggestion
            EntityReference::Slot(loc) => Expr::slot(slot).with_maybe_source_loc(loc.cloned()),
```

**Review by cdisselkoen — APPROVED** [review]
Looks good.

File: cedar-policy-core/Cargo.toml (hunk 6 above)
Comment (shaobo-he-aws) [reply cedar-policy-core/Cargo.toml:46]: It was added automatically by cargo. I will adjust it.

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**